File: Behavioral/Mediator/cpp/test_mediator.cpp
#include "mediator.hpp"
#include <gtest/gtest.h>
#include <sstream>
#include <iostream>

/**
 * @brief Fixture for testing the Mediator pattern.
 */
class MediatorTest : public ::testing::Test {
protected:
	std::ostringstream logStream; ///< Log stream for capturing output.

	void SetUp() override {
		// Redirect std::cout to capture output.
		std::cout.rdbuf(logStream.rdbuf());
	}

	void TearDown() override {
		// Restore std::cout.
		std::cout.rdbuf(nullptr);
	}
};

/**
 * @brief Test communication between colleagues using the mediator.
 */
TEST_F(MediatorTest, CommunicationBetweenColleagues) {
	ConcreteMediator mediator;

	ConcreteColleague colleague1(&mediator, "Colleague1");
	ConcreteColleague colleague2(&mediator, "Colleague2");

	mediator.addColleague(&colleague1);
	mediator.addColleague(&colleague2);

	colleague1.send("Hello from Colleague1");

	std::string output = logStream.str();
	EXPECT_NE(output.find("Colleague2 received message: Hello from Colleague1"), std::string::npos);
}

/**
 * @brief Test sending a message when no colleagues are registered.
 */
TEST_F(MediatorTest, SendMessageWithoutColleagues) {
	ConcreteMediator mediator;

	ConcreteColleague colleague1(&mediator, "Colleague1");

	colleague1.send("Message without other colleagues");

	std::string output = logStream.str();
	EXPECT_EQ(output, ""); // No output expected since no colleagues are registered to receive.
}

/**
 * @brief Test adding colleagues dynamically to the mediator.
 */
TEST_F(MediatorTest, DynamicColleagueAddition) {
	ConcreteMediator mediator;

	ConcreteColleague colleague1(&mediator, "Colleague1");
	ConcreteColleague colleague2(&mediator, "Colleague2");

	mediator.addColleague(&colleague1);
	colleague1.send("Hello before Colleague2 is added");

	mediator.addColleague(&colleague2);
	colleague1.send("Hello after Colleague2 is added");

	std::string output = logStream.str();
	EXPECT_EQ(output.find("Colleague2 received message: Hello before Colleague2 is added"), std::string::npos);
	EXPECT_NE(output.find("Colleague2 received message: Hello after Colleague2 is added"), std::string::npos);
}

File: Behavioral/Mediator/python/test_mediator.py
import pytest
from mediator import ConcreteMediator, ConcreteColleague


@pytest.fixture
def setup_mediator_and_colleagues():
	"""
	Fixture to set up a ConcreteMediator and two ConcreteColleagues.
	"""
	mediator = ConcreteMediator()
	colleague1 = ConcreteColleague(mediator, "Colleague1")
	colleague2 = ConcreteColleague(mediator, "Colleague2")

	mediator.add_colleague(colleague1)
	mediator.add_colleague(colleague2)

	return mediator, colleague1, colleague2


def test_communication_between_colleagues(caplog, setup_mediator_and_colleagues):
	"""
	Tests that colleagues can communicate with each other via the mediator.
	"""
	_, colleague1, colleague2 = setup_mediator_and_colleagues

	with caplog.at_level("INFO"):
		colleague1.send("Hello from Colleague1")

	assert "Colleague2 received message: Hello from Colleague1" in caplog.text


def test_send_message_without_other_colleagues(caplog):
	"""
	Tests that no messages are sent when there are no other colleagues.
	"""
	mediator = ConcreteMediator()
	colleague1 = ConcreteColleague(mediator, "Colleague1")
	mediator.add_colleague(colleague1)

	with caplog.at_level("INFO"):
		colleague1.send("Hello")

	assert "received message" not in caplog.text


def test_dynamic_colleague_addition(caplog):
	"""
	Tests dynamic addition of colleagues to the mediator.
	"""
	mediator = ConcreteMediator()
	colleague1 = ConcreteColleague(mediator, "Colleague1")
	mediator.add_colleague(colleague1)

	with caplog.at_level("INFO", logger="mediator"):
		colleague1.send("Hello before Colleague2 is added")

	colleague2 = ConcreteColleague(mediator, "Colleague2")
	mediator.add_colleague(colleague2)

	with caplog.at_level("INFO", logger="mediator"):
		colleague1.send("Hello after Colleague2 is added")

	assert "Colleague2 received message: Hello before Colleague2 is added" not in caplog.text
	assert "Colleague2 received message: Hello after Colleague2 is added" in caplog.text
